src/journal/journal.c

```diff
@@ -0,0 +1,612 @@
+        char *line = NULL;
+
+        fseek(fptr, 0, 0);
+        for(int i = 0; i < n; i++) {
+                if (getline(&line, &len, fptr) == -1) {
+                        rc = 1;
+                        break;
+                }
+        }
+        free(line);
+
+        return rc;
+}
+
+/**
+ * Copies contents of fptr to a known location.
+ *
+ * @param fptr A file pointer from source file.
+ *
+ * @returns 0 on success, 1 on failure
+ */
+static int copy_to_tmp(FILE *fptr)
+{
+        size_t len;
+        char *line = NULL;
+        FILE *fptr_tmp = NULL;
+
+        fptr_tmp = fopen(JOURNAL_TMP, "w");
+        if (!fptr_tmp) {
+                return 1;
+        }
+
+        // Copy lines to new file
+        while (getline(&line, &len, fptr) > -1) {
+                fprintf(fptr_tmp, "%s", line);
+        }
+
+        fclose(fptr_tmp);
+        free(line);
+
+        return 0;
+}
+
+/* Exported function */
+TelemJournal *open_journal(const char *journal_file)
+{
+        FILE *fptr = NULL;
+        char *boot_id = NULL;
+        struct TelemJournal *telem_journal;
+
+        // Use default location if journal_file parameter is NULL
+        fptr = (journal_file == NULL) ? fopen(JOURNAL_PATH, "a+") :
+               fopen(journal_file, "a+");
+
+        if (fptr == NULL) {
+#ifdef DEBUG
+                printf("Unable to open journal on disk\n");
+#endif
+                return NULL;
+        }
+
+        if (read_boot_id(&boot_id) != 0) {
+#ifdef DEBUG
+                printf("Failure reading boot_id\n");
+#endif
+                return NULL;
+        }
+
+        telem_journal = malloc(sizeof(struct TelemJournal));
+
+        telem_journal->fptr = fptr;
+        telem_journal->journal_file = \
+                (journal_file == NULL) ? strdup(JOURNAL_PATH) : strdup(journal_file);
+        telem_journal->boot_id = boot_id;
+        telem_journal->record_count = get_record_count(telem_journal);
+        telem_journal->record_count_limit = RECORD_LIMIT;
+
+#ifdef DEBUG
+        printf("Records in db: %d\n", telem_journal->record_count);
+#endif
+
+        return telem_journal;
+}
+
+/* Exported function */
+void close_journal(TelemJournal *telem_journal)
+{
+	if (telem_journal) {
+                free(telem_journal->boot_id);
+                free(telem_journal->journal_file);
+                fclose(telem_journal->fptr);
+                free(telem_journal);
+        }
+}
+
+/* Exported function */
+int print_journal(TelemJournal *telem_journal, char *classification,
+                   char *record_id, char *event_id, char *boot_id)
+{
+        int n = 0;
+        int count = 0;
+        char str_time[80] = {0};
+        char *line = NULL;
+        size_t len = 0;
+        FILE *journal_fileptr = NULL;
+        struct tm ts;
+        struct JournalEntry *entry = NULL;
+
+        if (telem_journal == NULL) {
+                return -1;
+        }
+
+        journal_fileptr = fopen(telem_journal->journal_file, "r");
```

Comment:
Should you be reusing `telem_journal->fptr` instead of opening `->journal_file` again, or closing `->fptr` at the end of `open_journal()`, or something else?